src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,250 @@
+import { readFile, rm, mkdir, readdir, writeFile } from "fs/promises";
+import path from "path";
+import { FlavorName } from "@ganache/flavors";
+
+export type DetachedInstance = {
+  instanceName: string;
+  pid: number;
+  startTime: number;
+  host: string;
+  port: number;
+  flavor: FlavorName;
+  cmd: string;
+  version: string;
+};
+
+const READY_MESSAGE = "ready";
+
+const START_ERROR = "An error ocurred spawning a detached instance of Ganache:";
+
+const dataPath = new Promise<string>(async (resolve, reject) => {
+  const path = envPaths(`Ganache/instances`).data;
+  // it seems strange to call mkdir blindly, but this is to avoid a race
+  // condition where the directory is modified between checking it's existence,
+  // and creating it see https://nodejs.org/api/fs.html#fsexistspath-callback
+  try {
+    await mkdir(path, { recursive: true });
+  } catch (err) {
+    if ((err as { code: string }).code === "EEXIST") {
+      resolve(path);
+    }
+    reject(path);
+  }
+  resolve(path);
+});
+
+async function getInstanceFilePath(instanceName: string): Promise<string> {
+  return path.join(await dataPath, instanceName);
+}
+
+/**
+ * Notify that the detached instance has started and is ready to receive requests.
+ */
+export function notifyDetachedInstanceReady() {
+  // in "detach" mode, the parent will wait until the "ready" message is
+  // received before disconnecting from the child process.
+  process.send(READY_MESSAGE);
+}
+
+/**
+ * Attempt to find and remove the instance file for a detached instance.
+ * @param  {string} instanceName the name of the instance to be removed
+ * @returns boolean indicating whether the instance file was cleaned up successfully
+ */
+export async function removeDetachedInstanceFile(
+  instanceName: string
+): Promise<boolean> {
+  const instanceFilename = await getInstanceFilePath(instanceName);
+  try {
+    await rm(instanceFilename);
+    return true;
+  } catch (err) {}
+  return false;
+}
+
+/**
+ * Attempts to stop a detached instance with the specified instance name by
+ * sending a SIGTERM signal. Returns a boolean indicating whether the process
+ * was found. If the PID is identified, but the process is not found, any
+ * corresponding instance file will be removed.
+ *
+ * Note: This does not guarantee that the instance actually stops.
+ * @param  {string} instanceName
+ * @returns boolean indicating whether the instance was found.
+ */
+export async function stopDetachedInstance(
+  instanceName: string
+): Promise<boolean> {
+  try {
+    // getDetachedInstanceByName() throws if the instance file is not found or
+    // cannot be parsed
+    const instance = await getDetachedInstanceByName(instanceName);
+
+    // process.kill() throws if the process was not found (or was a group
+    // process in Windows)
+    process.kill(instance.pid, "SIGTERM");
+  } catch (err) {
+    return false;
+  } finally {
+    removeDetachedInstanceFile(instanceName);
```

Comment:
should this be

```suggestion
    await removeDetachedInstanceFile(instanceName);
```
?

Reply:
Good catch